tasks/auditd.yml

```diff
@@ -18,41 +18,14 @@
 
 - name: set RedHat audit grub cmdline
   become: 'yes'
-  lineinfile:
-    line: 'GRUB_CMDLINE_LINUX="$GRUB_CMDLINE_LINUX {{ grub_audit_cmdline }} {{ grub_audit_backlog_cmdline }}"'
-    dest: /etc/default/grub
-    insertafter: '^GRUB_CMDLINE_LINUX="[a-z]+'
-    state: present
-    create: 'no'
-    mode: 0644
-    owner: root
-    group: root
+  command: grubby --update-kernel=ALL --args="{{ grub_audit_cmdline }} {{ grub_audit_backlog_cmdline }}"
```

Comment:
This is not a idempotent task, and needs a check so it isn't executed if `{{ grub_audit_cmdline }} {{ grub_audit_backlog_cmdline }}` is present in `GRUB_CMDLINE_LINUX`.

Reply:
Grubby grants options presence. If already present, then will be not added. If missing, grub will add. 

Additionally, in original task, the mentioned check is missing, isn't it ?  So it can be duplicated by old way. The new one with grubby grants no duplicity

Reply:
https://github.com/konstruktoid/ansible-role-hardening/blob/master/tasks/auditd.yml#L22 is idempotent since `lineinfile` wont change any file if the line is present.

that might be so, but `ansible` doesn't agree.

```shell
$ ansible-playbook -v -i '127.0.0.1,' -c local audit_grub.yml 
Using /etc/ansible/ansible.cfg as config file

PLAY [all] *************************************************************************************************************************************************************************************************

TASK [Gathering Facts] *************************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [set RedHat audit grub cmdline] ***********************************************************************************************************************************************************************
changed: [127.0.0.1] => {"changed": true, "cmd": ["grubby", "--update-kernel=ALL", "--args=audit_backlog_limit=8192 audit=1"], "delta": "0:00:00.177467", "end": "2021-02-16 13:01:23.498124", "rc": 0, "start": "2021-02-16 13:01:23.320657", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

PLAY RECAP *************************************************************************************************************************************************************************************************
127.0.0.1                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

$ ansible-playbook -v -i '127.0.0.1,' -c local audit_grub.yml 
Using /etc/ansible/ansible.cfg as config file

PLAY [all] *************************************************************************************************************************************************************************************************

TASK [Gathering Facts] *************************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [set RedHat audit grub cmdline] ***********************************************************************************************************************************************************************
changed: [127.0.0.1] => {"changed": true, "cmd": ["grubby", "--update-kernel=ALL", "--args=audit_backlog_limit=8192 audit=1"], "delta": "0:00:00.176944", "end": "2021-02-16 13:01:28.831339", "rc": 0, "start": "2021-02-16 13:01:28.654395", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

PLAY RECAP *************************************************************************************************************************************************************************************************
127.0.0.1                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 
```

Reply:
Oh I see. The grubby doesn't report if changed something or not.  But it add parameters only once. if already exists, it's not duplicit them. I have checked this again. 

The grubby is recommended way how to modify grub2 entries on RHEL now.... 

Provide an detection before the task can be really complicated (due possible several kernel configurations).

Is this behavior real problem ?

Reply:
Define a "real problem".

https://cumulusnetworks.com/blog/netdevops-important-idempotence/
https://molecule.readthedocs.io/en/latest/faq.html#why-is-my-idempotence-action-failing

Reply:
Thhat if the task report the change, even if all paremeters are really set. is it a problem? 

because if you use this:
line: 'GRUB_CMDLINE_LINUX="$GRUB_CMDLINE_LINUX {{ grub_audit_cmdline }} {{ grub_audit_backlog_cmdline }}"'

then by my opinion

1. If only one parameter is already set, then no change is provided
2. If another taks will provide something simmilar 'GRUB_CMDLINE_LINUX="$GRUB_CMDLINE_LINUX param1 param2" then previous change will be lost, because who upates the $GRUB_CMDLINE_LINUX variable? 

I think that lost changes or do not get them all is more serious then this one with change detection...

Reply:
`grubby`  is the correct way to update it on CentOS/RHEL, since we can change this on Ubuntu et al in the `grub.d` directory.
what is needed is that if `{{ grub_audit_cmdline }} {{ grub_audit_backlog_cmdline }}` is already present, `grubby` shouldn't be used. this is to ensure nothing is changed unless it has to be.
I know the file itself doesn't change, but as long as ansible considers otherwise, the role isn't idempotent.

this is a small issue, and I can fix it after merge.

Reply:
The problem with detection here is, that sometimes it can be in grubenv file, sometimes somewhere else, depends if EFI or BIOS is used, etc :(

maybe the  `grub2-editenv - list | grep kernelopts` can be used and then pres the output. But it's really lot of work for small detection.. :(